# Cvičenie 2 - Funkcie pre cyklické spracovanie, načítanie dát a práca s knižnicou <code>dplyr</code></h2>
Cieľom druhého cvičenia je osvojiť si použitie funkcií pre cyklické spracovanie, nacítať dáta z rôznych súborov a predspracovať ich pomocou knižnice <code>dplyr</code>.
        
# <span style="color:lightpink"> **Funkcie pre cyklické spracovanie** </span>

### 1. `apply()`

Funkcia `apply()` sa používa na evaluáciu funkcie v rámci ohraničení poľa. To znamená, že si vyberiem pole (často sú to riadky alebo stĺpce matice), na ktoré aplikujem zvolenú funkciu. 

Prvý uvedený príklad aplikuje funkciu `sum` na stpĺce a následne na riadky matice M. Spusti skript a sleduj čo sa udeje v druhom príklade.

In [1]:
M =matrix(1:6,3,2)
print(M)

     [,1] [,2]
[1,]    1    4
[2,]    2    5
[3,]    3    6


In [2]:
apply(M,2,sum) # sucet jednotlivých stlpcov (2)

[1]  6 15

In [3]:
apply(M,1,sum) # sucet jednotlivych riadkov (1)

[1] 5 7 9

In [4]:
M[3,2]

[1] 6

In [41]:
ncol(M)

[1] 2

In [5]:
for(i in 1:dim(M)[2]){
    suma = 0
    for(j in 1:dim(M)[1]){
        suma = suma + M[j,i]
    }
    print(suma)
}

[1] 6
[1] 15


In [6]:
N = matrix(c(1:10, 11:20), 10, 2)  
print(N)

      [,1] [,2]
 [1,]    1   11
 [2,]    2   12
 [3,]    3   13
 [4,]    4   14
 [5,]    5   15
 [6,]    6   16
 [7,]    7   17
 [8,]    8   18
 [9,]    9   19
[10,]   10   20


In [7]:
print(apply(N, 1:2, function(x) x/2))  

      [,1] [,2]
 [1,]  0.5  5.5
 [2,]  1.0  6.0
 [3,]  1.5  6.5
 [4,]  2.0  7.0
 [5,]  2.5  7.5
 [6,]  3.0  8.0
 [7,]  3.5  8.5
 [8,]  4.0  9.0
 [9,]  4.5  9.5
[10,]  5.0 10.0


In [8]:
apply(N, 1:2, function(x) x/2)  # porovnajte rozdiel vo vypise - plati len pre jupyter notebook

0.5,5.5
1.0,6.0
1.5,6.5
2.0,7.0
2.5,7.5
3.0,8.0
3.5,8.5
4.0,9.0
4.5,9.5
5.0,10.0


### 2. <code>lapply()</code>, <code>sapply()</code> 

Funkcia <code>lapply()</code> sa používa pri zoznamoch (`list`), pričom výsledok je opäť zoznam (`list`). Ak máme na vstupe iný dátový typ, môžeme ho jednoducho zmeniť prílazom <code>as.list()</code>. Funkcia <code>sapply()</code> sa snaží zjednodušiť výsledok <code>lapply()</code>. 
        
Výsledkom <code>sapply()</code> je:
        <ul>
          <li> vektor (ak elementy majú dĺžku 1),</li>
          <li> matica (ak elementy majú ROVNAKÚ dĺžku, väčšiu ako 1),</li>
          <li> nak vráti zoznam (`list`) ako funkcia <code>lapply()</code>. </li>
        </ul> </p>

In [9]:
# lapply vysledok list 
# sapply vysledok vektor/ matica (zjednoduseny lapply)
X = list(a = 1:5, b = 6:10)

In [10]:
X

$a
[1] 1 2 3 4 5

$b
[1]  6  7  8  9 10

In [11]:
lapply(X,mean)
sapply(X,mean) 

$a
[1] 3

$b
[1] 8

a b 
3 8

In [12]:
# poznamka: kvantil = hodnota pod ktorou lezi definovana cast udajov - rozmedzie je 0-1
#           kvantil 0,5 je median 
#           kvaRtil =  prvy kvartil 25%  = kvantil 0.25           
#                      druhy kvartil 50% = kvantil 0.5 = median
#                      treti kvartil 75% = kvantil 0.75

lapply(X,quantile)
sapply(X,quantile)

$a
  0%  25%  50%  75% 100% 
   1    2    3    4    5 

$b
  0%  25%  50%  75% 100% 
   6    7    8    9   10

,a,b
0%,1,6
25%,2,7
50%,3,8
75%,4,9
100%,5,10


In [13]:
lapply(X,quantile,c(0,0.4,1))
sapply(X,quantile,c(0,0.4,1))

$a
  0%  40% 100% 
 1.0  2.6  5.0 

$b
  0%  40% 100% 
 6.0  7.6 10.0

,a,b
0%,1.0,6.0
40%,2.6,7.6
100%,5.0,10.0


In [14]:
Y = list(a = 1, b = 1:3, c = 10:100) 
Y

$a
[1] 1

$b
[1] 1 2 3

$c
 [1]  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28
[20]  29  30  31  32  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47
[39]  48  49  50  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66
[58]  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85
[77]  86  87  88  89  90  91  92  93  94  95  96  97  98  99 100

In [15]:
lapply(Y,length)  
sapply(Y,length)  

$a
[1] 1

$b
[1] 3

$c
[1] 91

a  b  c 
 1  3 91

In [16]:
lapply(Y,sum)
sapply(Y,sum)

$a
[1] 1

$b
[1] 6

$c
[1] 5005

a    b    c 
   1    6 5005

### 3. <code>mapply()</code> 

Na základe nasledujúcich príkladov popíšte funkciu <code>mapply()</code>:

In [17]:
l1 = list(a = c(1:10), b = c(11:20))
l2 = list(c = c(21:30), d = c(31:40))

In [18]:
mapply(sum, l1$a, l1$b, l2$c, l2$d)

[1]  64  68  72  76  80  84  88  92  96 100

In [19]:
mapply(sum,1:5,1:5,1:5) #1,2,3,4,5   1,2,3,4,5    1,2,3,4,5

[1]  3  6  9 12 15

In [20]:
mapply(rep, 1:4, 6:9) #1,2,3,4     6,7,8,9

[[1]]
[1] 1 1 1 1 1 1

[[2]]
[1] 2 2 2 2 2 2 2

[[3]]
[1] 3 3 3 3 3 3 3 3

[[4]]
[1] 4 4 4 4 4 4 4 4 4

### 4. <code>tapply()</code>, <code>split()</code> 

Funkcia <code>tapply()</code> vypočíta zadanú funkciu pre každú faktorovú premennú vo vektore. 
Je to veľmi užitočná funkcia, ktorá vám umožňuje vytvoriť podmnožinu vektora a potom aplikovať niektoré funkcie na každú z podmnožín.
Vstupom je vektor a dátovy typ ktorý je alebo môže byť prevedený na faktor.

Funkcia <code>split()</code> rozdelí vstup do skupín podľa faktorov.

In [21]:
x = c(1,2,4,1,5,3,2,1)
x
y = c("ekonometria", "JDA", "ekonometria", "JDA", "SU", "JDA","SU","SU")
y

[1] 1 2 4 1 5 3 2 1

[1] "ekonometria" "JDA"         "ekonometria" "JDA"         "SU"         
[6] "JDA"         "SU"          "SU"

In [22]:
tapply(x, y, mean)

ekonometria         JDA          SU 
   2.500000    2.000000    2.666667

In [23]:
tapply(x,y,mean,simplify = FALSE)

$ekonometria
[1] 2.5

$JDA
[1] 2

$SU
[1] 2.666667

In [24]:
split(x,y)

$ekonometria
[1] 1 4

$JDA
[1] 2 1 3

$SU
[1] 5 2 1


# <span style="color:lightpink"> **Získavanie a načítavanie dát** </span>

### 1. Stiahnutie súborov z webu
Vo vašom pracovnom priečinku si vytvorte prieňčinok `DATA`. nastavte cestu do tohto priečinka a na základe nasledujúceho skriptu stiahnite dáta, ktoré pomenujeme `iris.csv`:

In [25]:
fileUrl = "http://people.tuke.sk/peter.butka/res/irisdata.csv"
download.file(fileUrl, destfile = "DATA/iris2.csv")
list.files("DATA")

[1] "camera.csv" "iris2.csv"

### 2. Načítanie lokálnych textových súborov
Textové súbory načítavame pomocou <code>read.table()</code>. Pre <samp>.csv</samp> súbory existuje špeciálny
typ funkcie <code>read.table()</code> a to  <code>read.csv()</code> a <code>read.csv2()</code>, pričom:
        <ul>
            <li><code>read.csv()</code> ma default separátor `,` a desatinnú čiarku `.` </li>
             <li><code>read.csv2()</code> ma default separátor `;` a desatinnú čiarku `,`</li>
        </ul>

Spusti nasledujúci skript a popíš funkcie <code>head()</code> a <code>tail()</code>.

In [26]:
iris = read.table("DATA/iris.csv", sep = ";",header = TRUE)

Warning message in file(file, "rt"):
“cannot open file 'DATA/iris.csv': No such file or directory”


ERROR: Error in file(file, "rt"): cannot open the connection


In [27]:
# OR
iris1 = read.csv("DATA/iris.csv")          # preco nefunguje? 

Warning message in file(file, "rt"):
“cannot open file 'DATA/iris.csv': No such file or directory”


ERROR: Error in file(file, "rt"): cannot open the connection


In [28]:
# OR
iris2 = read.csv2("DATA/iris.csv")

Warning message in file(file, "rt"):
“cannot open file 'DATA/iris.csv': No such file or directory”


ERROR: Error in file(file, "rt"): cannot open the connection


In [29]:
head(iris,20)
tail(iris)

,Sepal.Length,Sepal.Width,Petal.Length,Petal.Width,Species
,<dbl>,<dbl>,<dbl>,<dbl>,<fct>
1,5.1,3.5,1.4,0.2,setosa
2,4.9,3.0,1.4,0.2,setosa
3,4.7,3.2,1.3,0.2,setosa
4,4.6,3.1,1.5,0.2,setosa
5,5.0,3.6,1.4,0.2,setosa
6,5.4,3.9,1.7,0.4,setosa
7,4.6,3.4,1.4,0.3,setosa
8,5.0,3.4,1.5,0.2,setosa
9,4.4,2.9,1.4,0.2,setosa


,Sepal.Length,Sepal.Width,Petal.Length,Petal.Width,Species
,<dbl>,<dbl>,<dbl>,<dbl>,<fct>
145,6.7,3.3,5.7,2.5,virginica
146,6.7,3.0,5.2,2.3,virginica
147,6.3,2.5,5.0,1.9,virginica
148,6.5,3.0,5.2,2.0,virginica
149,6.2,3.4,5.4,2.3,virginica
150,5.9,3.0,5.1,1.8,virginica


### 3. JSON súborov

JSON = JavaScript Object Notation, ukladanie a zdieľaie dát v štruktúrovanej formw key:value.

Otvorte v prehliadači tento <a href="http://people.tuke.sk/peter.butka/res/data.JSON" > JSON súbor</a> a oboznámte sa s jeho štruktúrou.

Následne načítajte tento JSON súbor do R a vykonajte nasludujúce príkazy:

In [30]:
install.packages("curl")
library(jsonlite)

In [42]:
data_json = fromJSON("http://people.tuke.sk/peter.butka/res/data.JSON")   # nacitanie dat

In [43]:
names(data_json)                                                          # vypise names t.j. NAZVY stlpcov data_json 
names(data_json$adresa)                                                   # vypise names t.j. NAZVY, data_json$adresa t.j. stlpcov z ktorych sa sklada adresa

[1] "meno"       "priezvisko" "adresa"     "vek"

[1] "ulica" "cislo" "mesto"

In [44]:
data_json$adresa$mesto                                                    # vypise z tabulky, zo stplca ADRESA, vsetky MESTA

[1] "Kosice"     "Presov"     "Michalovce"

In [45]:
data_json$meno

[1] "Peter" "Jan"   "Jan"

  <h3>Úlohy</h3>
        <div class="alert alert-info">
          <h4>Úloha 2.1</h4>
          <p>
              Stihanite si dáta zo web stránky <samp>http://people.tuke.sk/viera.maslej.kresnakova/JDA/camera.csv,</samp>, uložte ich do priečinku <samp>DATA</samp> a pomenujte ich <samp>camera.csv</samp>
            </p>
        </div>


In [34]:
fileUrl = "http://people.tuke.sk/viera.maslej.kresnakova/JDA/camera.csv"
download.file(fileUrl, destfile = "DATA/camera.csv")
list.files("DATA")

[1] "camera.csv" "iris2.csv"

 <div class="alert alert-info">
            <h4>Úloha 2.2</h4>
            <p>
                Načítajte dáta <samp>camera</samp> ako tabuľku, <samp>csv</samp> alebo <samp>csv2</samp> a vypíšte prvých 5 riadkov.
            </p>
          </div>

In [35]:
camera = read.table("DATA/camera.csv", sep = ";",header = TRUE)
head(camera,5)
# OR
#iris2 = read.csv2("DATA/iris.csv")

,address.direction.street.crossStreet.intersection.Location.1.2010.Census.Neighborhoods.2010.Census.Wards.Precincts.Zip.Codes
,<chr>
1,"GARRISON BLVD & WABASH AVE,E/B,Garrison ,Wabash Ave,Garrison & Wabash Ave,(39.341209, -76.683117),252,63,27295"
2,"HILLEN ST & FORREST ST,W/B,Hillen ,Forrest St,Hillen & Forrest St,(39.29686, -76.605532),179,108,13645"
3,"EDMONDSON AVE & N ATHOL AVE,E/B,Edmonson ,Woodbridge Ave,Edmonson & Woodbridge Ave,(39.293453, -76.689391),213,75,27950"
4,"YORK RD & GITTINGS AVE,S/B,York Rd ,Gitting Ave,York Rd & Gitting Ave,(39.370493, -76.609812),37,270,14009"
5,"RUSSELL ST & W HAMBURG ST,S/B,Russell ,Hamburg St,Russell & Hamburg St,(39.279819, -76.623911),250,178,27953"


<div class="alert alert-info">
              <h4>Úloha 2.3</h4>
              <p>
                  Načítajte dáta z <samp>https://api.github.com/users/jtleek/repos</samp>, vypíšte zo stĺpca OWNER všetky LOGIN.
              </p>
          </div>

In [36]:
install.packages("curl")
library(jsonlite)
data_json = fromJSON("https://api.github.com/users/jtleek/repos")
data_json$owner$login   

Warning message in install.packages("curl"):
“installation of package ‘curl’ had non-zero exit status”
Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done



<h3>Úlohy na precvičenie</h3>
<div class="alert alert-success">
<h4>Úloha 2.4</h4>
  <p>
    Vytvorte maticu o rozmere 5 riadkov a 7 stĺpcov a naplňte ju párnymi číslami.
    Skúste pri tom využiť funkciu <samp>seq()</samp>
   <br>
    1. Vhodnou funkciou spočítajte priemernú hodnotu jednotlivých stĺpcov.
   <br>
    Výsledok:
      <pre>6 16 26 36 46 56 66</pre>
   <br>
    2. Na všetky prvky matice použite funkciu (x^2 - 1)
    <br>
    Výsledok:
  </p>
    <pre>
     [,1] [,2] [,3] [,4] [,5] [,6] [,7]
[1,]    3  143  483 1023 1763 2703 3843
[2,]   15  195  575 1155 1935 2915 4095
[3,]   35  255  675 1295 2115 3135 4355
[4,]   63  323  783 1443 2303 3363 4623
[5,]   99  399  899 1599 2499 3599 4899
    </pre>
</div>

In [ ]:
M = matrix(nrow = 5, ncol = 7)
M[] = seq(from=2, by=2, length.out = length(M))
M
colMeans(M)
N = M*M-1
N

In [ ]:
# vytvoríme maticu o rozmere 5 x 7
mat <- matrix(nrow = 5, ncol = 7)

# naplníme ju párnymi číslami pomocou seq()
mat[] <- seq(from = 0, by = 2, length.out = length(mat))

# vypíšeme maticu
mat

<div class="alert alert-success">
<h4>Úloha 2.5</h4>
  <p>
    Načítajte si dataset <samp>cars</samp> ktorý je súčasťou R napr. ako <samp>df=cars</samp>
    <br>
    1. Vhodnou funkciou nájdite minimálne a maximálne hodnoty pre oba stĺpce speed a distance.
    <br>
    Výsledok
    <pre>
<b>speed</b>   4 
<b>dist</b>    2
<b>speed</b>  25 
<b>dist</b>  120
    </pre>
    <br>
    2. Použite vlastnú funkciu avg (min(x)+max(x)/2) na získanie priemeru hodnôt zo stĺpcov speed a distance. 
    <br>
    Výsledok
    <pre>
<b>speed</b>  14.5
<b>dist</b>   61
    </pre>
      
  </p>
</div>

In [ ]:
df = cars
min(df$speed)
min(df$dist)
max(df$speed)
max(df$dist)

In [37]:
f = function(x){
  return( (min(x)+max(x)) /2 )   
}
 
f(df$speed)
f(df$dist)

ERROR: Error in df$speed: object of type 'closure' is not subsettable


<div class="alert alert-success">
<h4>Úloha 2.6</h4>
  <p>
    Použite dataset iris a funkciu tapply na získanie hodnoty medián zo stĺpca <samp>Petal.Width</samp> pre jednotlivé druhy kvetov, ktoré sú v stĺpci <samp>Species</samp>. 
Jednotlivé druhy (kategórie) si môžete zobraziť funkciou <samp>unique()</samp> pre zvolený stĺpec.
   <br>
    Výsledok
   </p>
    <pre>
<b>setosa</b>         0.2
<b>versicolor</b>     1.3
<b>virginica</b>      2
    </pre>
</div>

In [38]:
df = iris
#unique_species = unique(df$Species)
median = tapply(df$Petal.Width, df$Species, median)
median

setosa versicolor  virginica 
       0.2        1.3        2.0

<div class="alert alert-success">
<h4>Úloha 2.7</h4>
  <p>
    Skúste všetky predchádzajúce úlohy nahradiť vlastnými funkciamu použitím cyklusov <samp>for</samp>, <samp>while</samp> alebo <samp>repeat</samp>.
  </p>
</div>